File: app/src/main/java/com/bsoft/inventorymanager/repositories/PurchaseRepositoryImpl.java
package com.bsoft.inventorymanager.repositories;

import com.bsoft.inventorymanager.models.Product;
import com.bsoft.inventorymanager.models.Purchase;
import com.bsoft.inventorymanager.models.PurchaseItem;
import com.bsoft.inventorymanager.model.Supplier;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.WriteBatch;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class PurchaseRepositoryImpl implements PurchaseRepository {

    private final FirebaseFirestore db;

    @Inject
    public PurchaseRepositoryImpl(FirebaseFirestore db) {
        this.db = db;
    }

    @Override
    public void savePurchase(Purchase purchase, List<PurchaseItem> items, PurchaseCallback callback) {
        WriteBatch batch = db.batch();
        DocumentReference purchaseRef;

        if (purchase.getDocumentId() == null || purchase.getDocumentId().isEmpty()) {
            purchaseRef = db.collection("purchases").document();
            purchase.setDocumentId(purchaseRef.getId());
        } else {
            purchaseRef = db.collection("purchases").document(purchase.getDocumentId());
        }

        batch.set(purchaseRef, purchase); // Save the Purchase document

        // Update stock for each item (INCREMENT for purchases)
        for (PurchaseItem item : items) {
            if (item.getProductId() != null) {
                DocumentReference productRef = db.collection("products").document(item.getProductId());
                batch.update(productRef, "quantity", FieldValue.increment(item.getQuantity()));
            }
        }

        batch.commit()
                .addOnSuccessListener(aVoid -> callback.onSuccess(purchase.getDocumentId()))
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void updatePurchase(Purchase purchase, List<PurchaseItem> items, PurchaseCallback callback) {
        // Similar to updateSale, stock logic on update is complex.
        // For now, mirroring SaleRepositoryImpl approach of basic save, but
        // acknowledged complexity.
        // In a real scenario, we'd need to diff existing items to revert stock.

        WriteBatch batch = db.batch();
        DocumentReference purchaseRef = db.collection("purchases").document(purchase.getDocumentId());
        batch.set(purchaseRef, purchase);

        batch.commit()
                .addOnSuccessListener(aVoid -> callback.onSuccess(purchase.getDocumentId()))
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void getPurchase(String purchaseId, GetPurchaseCallback callback) {
        db.collection("purchases").document(purchaseId).get()
                .addOnSuccessListener(documentSnapshot -> {
                    if (documentSnapshot.exists()) {
                        Purchase purchase = documentSnapshot.toObject(Purchase.class);
                        if (purchase != null) {
                            purchase.setDocumentId(documentSnapshot.getId());
                            callback.onSuccess(purchase);
                        } else {
                            callback.onFailure(new Exception("Purchase data error"));
                        }
                    } else {
                        callback.onFailure(new Exception("Purchase not found"));
                    }
                })
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void getProductByBarcode(String barcode, ProductCallback callback) {
        db.collection("products")
                .whereEqualTo("productCode", barcode)
                .limit(1)
                .get()
                .addOnSuccessListener(querySnapshot -> {
                    if (!querySnapshot.isEmpty()) {
                        Product product = querySnapshot.getDocuments().get(0).toObject(Product.class);
                        if (product != null) {
                            product.setDocumentId(querySnapshot.getDocuments().get(0).getId());
                            callback.onSuccess(product);
                        } else {
                            callback.onFailure(new Exception("Product not found"));
                        }
                    } else {
                        callback.onFailure(new Exception("Product not found"));
                    }
                })
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void getProduct(String productId, ProductCallback callback) {
        db.collection("products").document(productId).get()
                .addOnSuccessListener(documentSnapshot -> {
                    if (documentSnapshot.exists()) {
                        Product product = documentSnapshot.toObject(Product.class);
                        if (product != null) {
                            product.setDocumentId(documentSnapshot.getId());
                            callback.onSuccess(product);
                        } else {
                            callback.onFailure(new Exception("Product data error"));
                        }
                    } else {
                        callback.onFailure(new Exception("Product not found"));
                    }
                })
                .addOnFailureListener(callback::onFailure);
    }

    @Override
    public void getSuppliers(SuppliersCallback callback) {
        db.collection("suppliers")
                .whereEqualTo("isActive", true)
                .get()
                .addOnSuccessListener(querySnapshot -> {
                    List<Supplier> suppliers = new java.util.ArrayList<>();
                    for (com.google.firebase.firestore.QueryDocumentSnapshot doc : querySnapshot) {
                        // Manual mapping to handle Timestamp -> Long conversion
                        Supplier s = new Supplier();
                        s.setDocumentId(doc.getId());
                        s.setName(doc.getString("name"));
                        s.setContactNumber(doc.getString("contactNumber"));
                        s.setAddress(doc.getString("address"));
                        s.setAge(doc.getLong("age") != null ? doc.getLong("age").intValue() : 0);
                        s.setPhoto(doc.getString("photo"));
                        s.setActive(Boolean.TRUE.equals(doc.getBoolean("isActive")));
                        s.setRating(doc.getDouble("rating") != null ? doc.getDouble("rating") : 0.0);
                        s.setPaymentTerms(doc.getString("paymentTerms"));
                        s.setLeadTime(doc.getLong("leadTime") != null ? doc.getLong("leadTime").intValue() : 0);
                        s.setPerformanceScore(
                                doc.getDouble("performanceScore") != null ? doc.getDouble("performanceScore") : 0.0);
                        s.setPreferredSupplier(Boolean.TRUE.equals(doc.getBoolean("preferredSupplier")));
                        s.setOutstandingPayment(
                                doc.getDouble("outstandingPayment") != null ? doc.getDouble("outstandingPayment")
                                        : 0.0);
                        s.setContractDetails(doc.getString("contractDetails"));
                        s.setProductsSupplied(doc.getString("productsSupplied"));

                        // Handle Timestamps safely (can be Timestamp or Long)
                        Object lastDeliveryObj = doc.get("lastDeliveryDate");
                        long lastDeliveryTime = 0;
                        if (lastDeliveryObj instanceof com.google.firebase.Timestamp) {
                            lastDeliveryTime = ((com.google.firebase.Timestamp) lastDeliveryObj).toDate().getTime();
                        } else if (lastDeliveryObj instanceof Number) {
                            lastDeliveryTime = ((Number) lastDeliveryObj).longValue();
                        }
                        s.setLastDeliveryDate(lastDeliveryTime);

                        s.setBankAccount(doc.getString("bankAccount"));
                        s.setTaxId(doc.getString("taxId"));

                        Object creationObj = doc.get("creationDate");
                        long creationTime = 0;
                        if (creationObj instanceof com.google.firebase.Timestamp) {
                            creationTime = ((com.google.firebase.Timestamp) creationObj).toDate().getTime();
                        } else if (creationObj instanceof Number) {
                            creationTime = ((Number) creationObj).longValue();
                        }
                        s.setCreationDate(creationTime);

                        s.setTotalSupplyAmount(
                                doc.getDouble("totalSupplyAmount") != null ? doc.getDouble("totalSupplyAmount") : 0.0);
                        s.setSupplyFrequency(
                                doc.getLong("supplyFrequency") != null ? doc.getLong("supplyFrequency").intValue() : 0);
                        s.setSupplierType(doc.getString("supplierType"));
                        s.setSupplierTier(doc.getString("supplierTier"));

                        suppliers.add(s);
                    }
                    callback.onSuccess(suppliers);
                })
                .addOnFailureListener(callback::onFailure);
    }
}

File: shared/src/androidMain/kotlin/com/bsoft/inventorymanager/repository/PurchaseRepositoryImpl.kt
package com.bsoft.inventorymanager.repository

import com.bsoft.inventorymanager.model.Product
import com.bsoft.inventorymanager.model.Purchase
import com.bsoft.inventorymanager.model.PurchaseItem
import com.bsoft.inventorymanager.model.Supplier
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QueryDocumentSnapshot
import kotlinx.coroutines.tasks.await
import java.util.Date

class PurchaseRepositoryImpl(private val db: FirebaseFirestore) : PurchaseRepository {

    override suspend fun getPurchase(purchaseId: String): Result<Purchase> {
        return try {
            val snapshot = db.collection("purchases").document(purchaseId).get().await()
            if (snapshot.exists()) {
                val data = snapshot.data ?: return Result.failure(Exception("No data"))
                val purchase = mapToPurchase(data, snapshot.id)
                Result.success(purchase)
            } else {
                Result.failure(Exception("Purchase not found"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun savePurchase(purchase: Purchase, items: List<PurchaseItem>): Result<String> {
        return try {
            val batch = db.batch()
            val purchaseRef = if (purchase.documentId.isNotEmpty()) {
                db.collection("purchases").document(purchase.documentId)
            } else {
                db.collection("purchases").document()
            }
            
            val finalPurchase = purchase.copy(documentId = purchaseRef.id, items = items)
            val purchaseMap = mapFromPurchase(finalPurchase)
            
            batch.set(purchaseRef, purchaseMap)

            // Update stock
            for (item in items) {
                if (item.productId.isNotEmpty()) {
                    val productRef = db.collection("products").document(item.productId)
                    batch.update(productRef, "quantity", FieldValue.increment(item.quantity.toLong()))
                }
            }
            
            batch.commit().await()
            Result.success(purchaseRef.id)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun updatePurchase(purchase: Purchase, items: List<PurchaseItem>): Result<Unit> {
        return try {
            val batch = db.batch()
            val purchaseRef = db.collection("purchases").document(purchase.documentId)
            val purchaseMap = mapFromPurchase(purchase.copy(items = items))
            
            batch.set(purchaseRef, purchaseMap)
            // Stock logic omitted for simplicity as per original
            
            batch.commit().await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun getProductByBarcode(barcode: String): Result<Product?> {
        return try {
            val query = db.collection("products")
                .whereEqualTo("productCode", barcode)
                .limit(1)
                .get()
                .await()
            
            if (!query.isEmpty) {
                val doc = query.documents[0]
                val product = mapToProduct(doc.data!!, doc.id)
                Result.success(product)
            } else {
                Result.success(null)
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun getProduct(productId: String): Result<Product> {
        return try {
            val snapshot = db.collection("products").document(productId).get().await()
            if (snapshot.exists()) {
                val product = mapToProduct(snapshot.data!!, snapshot.id)
                Result.success(product)
            } else {
                Result.failure(Exception("Product not found"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun getSuppliers(): Result<List<Supplier>> {
        return try {
            val query = db.collection("suppliers")
                .whereEqualTo("isActive", true)
                .get()
                .await()
            
            val suppliers = query.map { doc ->
                mapToSupplier(doc.data, doc.id)
            }
            Result.success(suppliers)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // --- Mappers ---

    private fun mapToPurchase(data: Map<String, Any>, id: String): Purchase {
        return Purchase(
            documentId = id,
            supplierId = data["supplierId"] as? String ?: "",
            supplierName = data["supplierName"] as? String ?: "",
            purchaseDate = (data["purchaseDate"] as? Timestamp)?.toDate()?.time ?: 0L,
            totalAmount = (data["totalAmount"] as? Number)?.toDouble() ?: 0.0,
            items = (data["items"] as? List<Map<String, Any>>)?.map { mapToPurchaseItem(it) } ?: emptyList()
            // Add other fields mapping
        )
    }
    
    private fun mapFromPurchase(purchase: Purchase): Map<String, Any?> {
        return mapOf(
            "documentId" to purchase.documentId,
            "supplierId" to purchase.supplierId,
            "supplierName" to purchase.supplierName,
            "purchaseDate" to Timestamp(Date(purchase.purchaseDate)),
            "totalAmount" to purchase.totalAmount,
            "items" to purchase.items.map { mapFromPurchaseItem(it) }
            // Add other fields
        )
    }

    private fun mapToPurchaseItem(data: Map<String, Any>): PurchaseItem {
        return PurchaseItem(
            productId = data["productId"] as? String ?: "",
            productName = data["productName"] as? String ?: "",
            quantity = (data["quantity"] as? Number)?.toInt() ?: 0,
            pricePerItem = (data["pricePerItem"] as? Number)?.toDouble() ?: 0.0,
            expiryDate = (data["expiryDate"] as? Timestamp)?.toDate()?.time
            // Other fields
        )
    }

    private fun mapFromPurchaseItem(item: PurchaseItem): Map<String, Any?> {
        return mapOf(
            "productId" to item.productId,
            "productName" to item.productName,
            "quantity" to item.quantity,
            "pricePerItem" to item.pricePerItem,
            "expiryDate" to item.expiryDate?.let { Timestamp(Date(it)) }
        )
    }
    
    private fun mapToProduct(data: Map<String, Any>, id: String): Product {
        return Product(
            documentId = id,
            name = data["name"] as? String ?: "",
            barcode = data["barcode"] as? String ?: "",
            quantity = (data["quantity"] as? Number)?.toInt() ?: 0,
            productCode = data["productCode"] as? String ?: ""
            // Other fields
        )
    }

    private fun mapToSupplier(data: Map<String, Any>, id: String): Supplier {
        return Supplier(
            documentId = id,
            name = data["name"] as? String ?: ""
            // Other fields
        )
    }
}
